src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java

```diff
@@ -248,17 +297,40 @@ private static void copyFile(UnixPath source,
             // set to true when file and attributes copied
             boolean complete = false;
             try {
-                // transfer bytes to target file
-                try {
+                boolean copied = false;
+                if (!directCopyNotSupported) {
+                    // copy bytes to target using platform function
                     long comp = Blocker.begin();
                     try {
-                        transfer(fo, fi, addressToPollForCancel);
+                        int res = directCopy0(fo, fi, addressToPollForCancel);
+                        if (res == 0) {
+                            copied = true;
+                        } else if (res == IOStatus.UNSUPPORTED) {
+                            directCopyNotSupported = true;
+                        }
+                    } catch (UnixException x) {
+                        x.rethrowAsIOException(source, target);
                     } finally {
                         Blocker.end(comp);
                     }
-                } catch (UnixException x) {
-                    x.rethrowAsIOException(source, target);
                 }
+
+                if (!copied) {
+                    // copy bytes to target via a temporary direct buffer
+                    int bufferSize = temporaryBufferSize(source, target);
+                    ByteBuffer buf = Util.getTemporaryDirectBuffer(bufferSize);
+                    long comp = Blocker.begin();
+                    try {
+                        bufferCopy0(fo, fi, ((DirectBuffer)buf).address(),
+                                    bufferSize, addressToPollForCancel);
+                    } catch (UnixException x) {
+                        x.rethrowAsIOException(source, target);
+                    } finally {
+                        Util.releaseTemporaryDirectBuffer(buf);
+                        Blocker.end(comp);
```

Comment:
It might be better to use a nested try-finally to ensure that the buffer get/release and blocker begin/end are symmetric.

Reply:
So changed in 6c7ac10c4a98e0d11cb487dfe816f460f7381002.